File: java/src/main/java/com/myriad/wdl/model/expressions/WdlUnaryOperation.java
package com.myriad.wdl.model.expressions;

import lombok.Getter;
import lombok.Setter;

/** Prefix unary operator expression. */
public final class WdlUnaryOperation implements WdlExpression {
  public static enum Operator {
    NOT("!"),
    NEGATIVE("-");

    @Getter private final String wdlString;

    private Operator(String wdlString) {
      this.wdlString = wdlString;
    }
  }

  @Getter @Setter private Operator operator;
  @Getter @Setter private WdlExpression operand;

  public WdlUnaryOperation() {}

  public WdlUnaryOperation(Operator operator, WdlExpression operandExpression) {
    setOperator(operator);
    setOperand(operandExpression);
  }

  @Override
  public String toString() {
    return getClass().getSimpleName();
  }

  @Override
  public ComponentType componentType() {
    return ComponentType.UNARY_OP;
  }
}

File: typescript/src/model/expressions/wdl-unary-operation.ts
/** Unary operation expression nodes and operator kinds. */
import { WdlExpressionComponentType, type WdlExpression } from './wdl-expression.js';

/** Supported unary operators from the WDL expression grammar. */
export enum WdlUnaryOperator {
  NOT = '!',
  NEGATIVE = '-',
}

/** Models a unary expression such as `!x` or `-n`. */
export class WdlUnaryOperation implements WdlExpression {
  /** Creates a unary expression from its operator and operand. */
  public constructor(
    private operatorValue?: WdlUnaryOperator,
    private operandValue?: WdlExpression,
  ) {}

  /** Returns the operator for the expression. */
  public getOperator(): WdlUnaryOperator | undefined {
    return this.operatorValue;
  }
  /** Sets the operator for the expression. */
  public setOperator(operator: WdlUnaryOperator | undefined): void {
    this.operatorValue = operator;
  }
  /** Returns the operand expression. */
  public getOperand(): WdlExpression | undefined {
    return this.operandValue;
  }
  /** Sets the operand expression. */
  public setOperand(operand: WdlExpression | undefined): void {
    this.operandValue = operand;
  }

  /** Returns the broad expression family for this node. */
  public componentType(): WdlExpressionComponentType {
    return WdlExpressionComponentType.UNARY_OP;
  }
}
